File: app/src/main/java/com/iamkatrechko/clipboardmanager/view/activity/MainActivity.java
package com.iamkatrechko.clipboardmanager.view.activity;

import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.widget.ExpandableListView;

import com.iamkatrechko.clipboardmanager.R;
import com.iamkatrechko.clipboardmanager.data.database.ClipboardDatabaseHelper.CategoryCursor;
import com.iamkatrechko.clipboardmanager.data.database.DatabaseDescription;
import com.iamkatrechko.clipboardmanager.view.adapter.navigation.NavGroups;
import com.iamkatrechko.clipboardmanager.view.adapter.navigation.NavigationMenuAdapter;
import com.iamkatrechko.clipboardmanager.view.fragment.ClipsListFragment;

/**
 * Основаня активность экрана со списком заметок
 * @author iamkatrechko
 *         Date: 01.11.2016
 */
public class MainActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<Cursor> {

    /** Тег для логирования */
    private static final String TAG = MainActivity.class.getSimpleName();

    /** Идентификатор загрузчика категорий */
    private static final int CATEGORIES_LOADER = 1;

    /** Виджет бокового меню */
    private DrawerLayout drawerLayout;

    /** Виджет списка бокового меню */
    private ExpandableListView listView;
    /** Адаптер бокового меню */
    private NavigationMenuAdapter adapter;
    /** Курсор со списком категорий */
    private CategoryCursor cursor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_layout);
        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));

        FragmentManager fragmentManager = getSupportFragmentManager();
        if (fragmentManager.getFragments().isEmpty()) {
            fragmentManager.beginTransaction()
                    .replace(R.id.container, ClipsListFragment.newInstance())
                    .commit();
        }

        findViewById(R.id.fab).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ClipsListFragment clipsFragment = (ClipsListFragment) getSupportFragmentManager().findFragmentById(R.id.container);
                clipsFragment.addNewClip();
            }
        });

        initNavigationView();


        getSupportLoaderManager().initLoader(CATEGORIES_LOADER, null, this);
    }

    @Override
    public void onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START);
        } else {
            ClipsListFragment clipsFragment = (ClipsListFragment) getSupportFragmentManager().findFragmentById(R.id.container);
            clipsFragment.onBackPressed();
        }
    }

    private void initNavigationView() {
        drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        listView = (ExpandableListView) findViewById(R.id.list_view_navigation);

        ActionBarDrawerToggle toogle = new ActionBarDrawerToggle(this,
                drawerLayout,
                (Toolbar) findViewById(R.id.toolbar),
                R.string.app_name,
                R.string.app_name);
        drawerLayout.setDrawerListener(toogle);
        toogle.syncState();

        adapter = new NavigationMenuAdapter();
        listView.setAdapter(adapter);

        /*listView.setItemChecked(NAV_MENU_POS_PROJECTS + 1, true);*/

        listView.setOnGroupClickListener(new ExpandableListView.OnGroupClickListener() {
            @Override
            public boolean onGroupClick(ExpandableListView expandableListView, View view, int groupPos, long groupId) {
                NavGroups group = NavGroups.values()[groupPos];
                switch (group) {
                    case CATEGORIES:
                        /*getFragment(NAV_MENU_POS_TODAY);*/
                        break;
                    /*case NAV_MENU_POS_WEEK:
                        getFragment(NAV_MENU_POS_WEEK);
                        break;*/
                    case SETTINGS:
                        startActivity(new Intent(getApplicationContext(), SettingsActivity.class));
                        break;
                }
                return false;
            }
        });

        listView.setOnChildClickListener(new ExpandableListView.OnChildClickListener() {
            @Override
            public boolean onChildClick(ExpandableListView expandableListView, View view, int groupPos, int childPos, long l) {
                NavGroups group = NavGroups.values()[groupPos];
                int index = expandableListView.getFlatListPosition(ExpandableListView.getPackedPositionForChild(groupPos, childPos));
                expandableListView.setItemChecked(index, true);
                switch (group) {
                    case CATEGORIES:
                        if (childPos > cursor.getCount() - 1) {
                            // Если нажата кнопка настроек групп
                            startActivity(new Intent(getApplicationContext(), EditCategoriesActivity.class));
                            return true;
                        }
                        ClipsListFragment clipsFragment = (ClipsListFragment) getSupportFragmentManager().findFragmentById(R.id.container);
                        cursor.moveToPosition(childPos);
                        clipsFragment.showClipsByCategoryId(cursor.getID());
                        drawerLayout.closeDrawer(GravityCompat.START);
                        return true;
                }
                return true;
            }
        });
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        switch (id) {
            case CATEGORIES_LOADER:
                Log.d(TAG, "onCreateLoader");
                return new CursorLoader(this,
                        DatabaseDescription.Category.CONTENT_URI, // Uri таблицы contacts
                        null, // все столбцы
                        null, // все записи
                        null, // без аргументов
                        null); // сортировка
            default:
                return null;
        }
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        Log.d(TAG, "onLoadFinished");
        cursor = new CategoryCursor(data);
        for (int i = 0; i < data.getCount(); i++) {
            cursor.moveToPosition(i);
        }
        adapter.setOfChildren(cursor);
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/com/iamkatrechko/clipboardmanager/view/activity/MainActivity.kt
package com.iamkatrechko.clipboardmanager.view.activity

import android.content.Intent
import android.os.Bundle
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.View
import android.widget.ExpandableListView
import com.iamkatrechko.clipboardmanager.R
import com.iamkatrechko.clipboardmanager.data.model.Category
import com.iamkatrechko.clipboardmanager.domain.loader.callback.CategoriesLoaderCallback
import com.iamkatrechko.clipboardmanager.view.adapter.NavigationMenuAdapter
import com.iamkatrechko.clipboardmanager.view.adapter.navigation.NavGroups
import com.iamkatrechko.clipboardmanager.view.fragment.ClipsListFragment

/**
 * Основаня активность экрана со списком заметок
 * @author iamkatrechko
 *         Date: 01.11.2016
 */
class MainActivity : AppCompatActivity() {

    /** Виджет бокового меню  */
    private lateinit var drawerLayout: DrawerLayout
    /** Виджет списка бокового меню  */
    private lateinit var listView: ExpandableListView

    /** Адаптер бокового меню  */
    private var adapter = NavigationMenuAdapter()
    /** Список категорий  */
    private val categories = ArrayList<Category>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_layout)
        setSupportActionBar(findViewById(R.id.toolbar))

        drawerLayout = findViewById<DrawerLayout>(R.id.drawer_layout)
        listView = findViewById<ExpandableListView>(R.id.list_view_navigation)

        val fragmentManager = supportFragmentManager
        if (fragmentManager.fragments.isEmpty()) {
            fragmentManager.beginTransaction()
                    .replace(R.id.container, ClipsListFragment.newInstance(1L))
                    .commit()
        }

        findViewById<View>(R.id.fab).setOnClickListener {
            val i = ClipEditActivity.newIntent(this, null)
            startActivity(i)
        }
        initNavigationView()

        supportLoaderManager.initLoader(CategoriesLoaderCallback.CATEGORIES_LOADER, null, CategoriesLoaderCallback(this, { categories ->
            this.categories.clear()
            this.categories.addAll(categories)
            adapter.setOfChildren(categories)
        }))
    }

    override fun onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            val clipsFragment = supportFragmentManager.findFragmentById(R.id.container) as ClipsListFragment
            clipsFragment.onBackPressed()
        }
    }

    private fun initNavigationView() {
        val toogle = ActionBarDrawerToggle(this,
                drawerLayout,
                findViewById<Toolbar>(R.id.toolbar),
                R.string.app_name,
                R.string.app_name)
        drawerLayout.setDrawerListener(toogle)
        toogle.syncState()

        listView.setAdapter(adapter)

        //listView.setItemChecked(0, true)

        listView.setOnGroupClickListener { expandableListView, view, groupPos, groupId ->
            val group = NavGroups.values()[groupPos]
            when (group) {
                NavGroups.ALL_CLIPS -> showByCategory(null)
                NavGroups.SETTINGS -> startActivity(Intent(applicationContext, SettingsActivity::class.java))
            }
            false
        }

        listView.setOnChildClickListener(ExpandableListView.OnChildClickListener { expandableListView, view, groupPos, childPos, l ->
            val group = NavGroups.values()[groupPos]
            val index = expandableListView.getFlatListPosition(ExpandableListView.getPackedPositionForChild(groupPos, childPos))
            expandableListView.setItemChecked(index, true)
            when (group) {
                NavGroups.CATEGORIES -> {
                    if (childPos > categories.size - 1) {
                        // Если нажата кнопка настроек групп
                        startActivity(Intent(applicationContext, EditCategoriesActivity::class.java))
                        return@OnChildClickListener true
                    }
                    showByCategory(categories[childPos].id)
                    return@OnChildClickListener true
                }
            }
            true
        })
    }

    /**
     * Отобржает список записей по категории
     * @param categoryId идентификатор категории
     */
    private fun showByCategory(categoryId: Long?) {
        val clipsFragment = supportFragmentManager.findFragmentById(R.id.container) as ClipsListFragment
        clipsFragment.showClipsByCategoryId(categoryId)
        drawerLayout.closeDrawer(GravityCompat.START)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
    }

    companion object {

        /** Тег для логирования  */
        private val TAG = MainActivity::class.java.simpleName

    }
}
